subxt/src/storage/storage_address.rs

```diff
@@ -54,64 +58,314 @@ pub struct Yes;
 /// via the `subxt` macro) or dynamic values via [`dynamic`].
 #[derive(Derivative)]
 #[derivative(
-    Clone(bound = "StorageKey: Clone"),
-    Debug(bound = "StorageKey: std::fmt::Debug"),
-    Eq(bound = "StorageKey: std::cmp::Eq"),
-    Ord(bound = "StorageKey: std::cmp::Ord"),
-    PartialEq(bound = "StorageKey: std::cmp::PartialEq"),
-    PartialOrd(bound = "StorageKey: std::cmp::PartialOrd")
+    Clone(bound = "Keys: Clone"),
+    Debug(bound = "Keys: std::fmt::Debug"),
+    Eq(bound = "Keys: std::cmp::Eq"),
+    Ord(bound = "Keys: std::cmp::Ord"),
+    PartialEq(bound = "Keys: std::cmp::PartialEq"),
+    PartialOrd(bound = "Keys: std::cmp::PartialOrd")
 )]
-pub struct Address<StorageKey, ReturnTy, Fetchable, Defaultable, Iterable> {
+pub struct Address<Keys: StorageMultiKey, ReturnTy, Fetchable, Defaultable, Iterable> {
     pallet_name: Cow<'static, str>,
     entry_name: Cow<'static, str>,
-    storage_entry_keys: Vec<StorageKey>,
+    keys: Keys,
     validation_hash: Option<[u8; 32]>,
     _marker: std::marker::PhantomData<(ReturnTy, Fetchable, Defaultable, Iterable)>,
 }
 
+/// A storage key, mostly used for static encoded values.
+/// The original value is only given during construction, but can be
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""), Debug(bound = ""))]
+pub struct StorageKey<K: ?Sized> {
+    bytes: Static<Encoded>,
+    _marker: std::marker::PhantomData<K>,
+}
+
+impl<K: codec::Encode + ?Sized> StorageKey<K> {
+    /// Creates a new static storage key
+    pub fn new(key: &K) -> Self {
+        StorageKey {
+            bytes: Static(Encoded(key.encode())),
+            _marker: std::marker::PhantomData,
+        }
+    }
+
+    /// Returns the scale-encoded bytes that make up this key
+    pub fn bytes(&self) -> &[u8] {
+        &self.bytes.0 .0
+    }
+}
+
+/// This trait should be implemented by anything that can be used as one or multiple storage keys.
+pub trait StorageMultiKey {
+    /// Iterator over the storage keys, each key implements EncodeAsType to
+    /// give the corresponding bytes to a `StorageHasher`.
+    fn keys_iter(&self) -> impl ExactSizeIterator<Item = &dyn EncodeAsType>;
```

Comment:
Ah ok, so while `&dyn EncodeAsType` won't work any more when I make `EncodeAsType` non-object-safe (ie generic over TypeResolvers), since Subxt will probably continue to use concrete types like `PortableRegistry` instead of relying too much on `TypeResolver`, it'll be easy enough to make a wrapper trait that is object safe in the same way I had to in `scale-encode` :)